### DISCUSSION TITLE
De-Fi Lesson 3 , testUsdGetValue , EvmError

### DISCUSSION THREAD

**Discussion by AllenGeorge08** [Cyfrin/foundry-full-course-cu#1975] [Q&A] [answered]
While running the testGetUsdValue() , I am getting this errror

```
allen@LN08:~/foundry/foundry-defi-stablecoin-f23$ forge test --mt testGetUsdValue -vvvv
[⠊] Compiling...
No files changed, compilation skipped

Ran 1 test for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: setup failed: EvmError: Revert] setUp() (gas: 0)
Traces:
  [9744808] DSCEngineTest::setUp()
    ├─ [4427227] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 22000 bytes of code
    ├─ [5264686] DeployDSC::run()
    │   ├─ [3746933] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372455] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [658896] → new ERC20Mock@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2829 bytes of code
    │   │   ├─ [372455] → new MockV3Aggregator@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1083 bytes of code
    │   │   ├─ [658896] → new ERC20Mock@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 100000000000 [1e11])
    │   │   │   └─ ← [Return] 2829 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 6970 bytes of code
    │   ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [750023] → new DecentralizedStableCoin@0x50EEf481cae4250d252Ae577A09bF514f224C6C4
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   └─ ← [Return] 3402 bytes of code
    │   ├─ [527102] → new DSCEngine@0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1
    │   │   └─ ← [Return] 1955 bytes of code
    │   ├─ [2446] DecentralizedStableCoin::transferOwnership(DSCEngine: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])
    │   │   ├─ emit OwnershipTransferred(previousOwner: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38], newOwner: DSCEngine: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x0000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000
    ├─ [0] 0x0000000000000000000000000000000000000000::activeNetworkConfig() [staticcall]
    │   └─ ← [Stop] 
    └─ ← [Revert] EvmError: Revert

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 822.23µs (0.00ns CPU time)

Ran 1 test suite in 1.20s (822.23µs CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: setup failed: EvmError: Revert] setUp() (gas: 0)

Encountered a total of 1 failing tests, 0 tests succeeded
allen@LN08:~/foundry/foundry-defi-stablecoin-f23$ 
```

Tried the solutions mentioned in [This solution in the discussions](https://github.com/Cyfrin/foundry-full-course-cu/discussions/1717)  and ran foundryup too...After running foundryup my console.log is also throwing errors 😂...

Below are my code....DCSEngineTest.t.sol 

```

// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity 0.8.22;

import {Test} from "forge-std/Test.sol";
import {console2} from "lib/forge-std/src/console2.sol";
import {DeployDSC} from "script/DeployDSC.s.sol";
import {DecentralizedStableCoin} from "src/DecentralizedStableCoin.sol";
import {DSCEngine} from "src/DSCEngine.sol";
import {HelperConfig} from "script/HelperConfig.s.sol";

contract DSCEngineTest is Test {
    DeployDSC deployer;
    DecentralizedStableCoin dsc;
    DSCEngine dsce;
    HelperConfig config;
    address ethUsdPriceFeed;
    address weth;

    function setUp() public {
        deployer = new DeployDSC();
        (dsc, dsce, config) = deployer.run();
        (ethUsdPriceFeed,, weth,,) = config.activeNetworkConfig();
    }

    /*//////////////////////////////////////////////////////////////
                              PRICE TESTS
    //////////////////////////////////////////////////////////////*/

    function testGetUsdValue() public {
        uint256 ethAmount = 15e18;
        //15e18 = 2000/ETH = 30000e18 Because 1 ETH equals $2000
        uint256 expectedUsd = 30000e18;
        uint256 actualUsd = dsce.getUsdInValue(weth, ethAmount);
        assertEq(expectedUsd, actualUsd);
    }
}
```

DeployDsc.s.sol

/```
/ SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity 0.8.22;

import {Script} from "forge-std/Script.sol";
import {DecentralizedStableCoin} from "src/DecentralizedStableCoin.sol";
import {DSCEngine} from "src/DSCEngine.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployDSC is Script {
    address[] public tokenAddresses;
    address[] public priceFeedAddresses;

    function run() external returns (DecentralizedStableCoin, DSCEngine, HelperConfig) {
        HelperConfig config = new HelperConfig();
        (address wethUsdPriceFeed, address wbtcUsdPriceFeed, address weth, address wbtc, uint256 deployerKey) =
            config.activeNetworkConfig();

        tokenAddresses = [weth, wbtc];
        priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];
        vm.startBroadcast();
        DecentralizedStableCoin dsc = new DecentralizedStableCoin();
        DSCEngine engine = new DSCEngine(tokenAddresses, priceFeedAddresses, address(dsc));
        dsc.transferOwnership(address(engine));
        vm.stopBroadcast();
    }
}
```


HelperConfig.s.sol


```
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity 0.8.22;

import {Script} from "forge-std/Script.sol";
import {MockV3Aggregator} from "test/mocks/MockV3Aggregator.sol";
import {ERC20Mock} from "@openzeppelin/contracts/mocks/ERC20Mock.sol";

contract HelperConfig is Script {
    uint8 public constant DECIMALS = 8;
    int256 public constant ETH_USD_PRICE = 2000e8;
    int256 public constant BTC_PRICE_FEED = 1000e8;
    NetworkConfig public activeNetworkConfig;

    struct NetworkConfig {
        address wethUsdPriceFeed; //ERC20 Version of Ethereum
        address wbtcUsdPriceFeed;
        address weth;
        address wbtc;
        uint256 deployerKey;
    }

    uint256 public DEFAULT_ANVIL_KEY = 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;

    constructor() {
        if (block.chainid == 11155111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else {
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public view returns (NetworkConfig memory) {
        return NetworkConfig({
            wethUsdPriceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306,
            wbtcUsdPriceFeed: 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43,
            weth: 0x72AFAECF99C9d9C8215fF44C77B94B99C28741e8,
            wbtc: 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063,
            deployerKey: vm.envUint("PRIVATE_KEY")
        });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (activeNetworkConfig.wethUsdPriceFeed != address(0)) {
            return activeNetworkConfig;
        }

        vm.startBroadcast();
        MockV3Aggregator ethUsdPriceFeed = new MockV3Aggregator(DECIMALS, ETH_USD_PRICE);
        ERC20Mock wethMock = new ERC20Mock("WETH", "WETH", msg.sender, 1000e8);

        MockV3Aggregator btcUsdPriceFeed = new MockV3Aggregator(DECIMALS, BTC_PRICE_FEED);
        ERC20Mock wbtcMock = new ERC20Mock("WBTC", "WBTC", msg.sender, 1000e8);

        vm.stopBroadcast();

        return NetworkConfig({
            wethUsdPriceFeed: address(ethUsdPriceFeed),
            wbtcUsdPriceFeed: address(btcUsdPriceFeed),
            weth: address(wethMock),
            wbtc: address(wbtcMock),
            deployerKey: DEFAULT_ANVIL_KEY
        });
    }
}

```

**Comment by EngrPips** [comment]
Hello @AllenOps08, Please ensure that you are calling functions at the appropriate address. The below shows that you are calling `activeNetworkConfig` on address(0). Why is that? 

```js
└─ ← [Return] 0x0000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000
    ├─ [0] 0x0000000000000000000000000000000000000000::activeNetworkConfig() [staticcall]
    │   └─ ← [Stop] 
    └─ ← [Revert] EvmError: Revert
```

**Reply by AllenGeorge08** [reply, marked as answer]
Thankss @EngrPips  bro...The problem was in DeployerDSC ..I was calling vm.broadcast() with an empty value in the run() function . Also I wasn't returning any values out of it...

**Reply by EngrPips** [reply]
Awesome, I am glad you figured it out. Keep crushing it, Friend.
